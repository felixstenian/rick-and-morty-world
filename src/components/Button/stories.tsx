import { Meta, Story } from '@storybook/react'
import Button from '.'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />

Default.args = {
  children: 'Click me'
}
