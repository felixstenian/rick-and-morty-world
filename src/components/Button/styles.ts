import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
  `,
  fullWidth: () => css`
    width: 100%;
  `
}

export const Wrapper = styled.button<ButtonProps>`
  border: 0;
  cursor: pointer;
  width: 20rem;

  ${({ theme, size, fullWidth }) => css`
    background: ${theme.colors.gradients.button};
    color: ${theme.colors.darkGray};
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};

    &:hover {
      background: ${theme.colors.green[200]};
    }

    ${!!size && wrapperModifiers[size](theme)};
    ${!!fullWidth && wrapperModifiers.fullWidth};
  `}
`
