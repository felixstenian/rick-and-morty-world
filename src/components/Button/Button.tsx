import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as Style from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  size = 'medium',
  fullWidth = false,
  ...props
}: ButtonProps) => (
  <Style.Wrapper size={size} fullWidth={fullWidth} {...props}>
    {children}
  </Style.Wrapper>
)

export default Button
