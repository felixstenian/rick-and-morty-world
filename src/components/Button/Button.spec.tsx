import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils'

import Button from '.'

describe('<Button />', () => {
  it('1. should render the reading', () => {
    const { container } = renderWithTheme(<Button>Click me</Button>)

    expect(
      screen.getByRole('button', { name: /Click me/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('2. should render the small size', () => {
    const { getByText } = renderWithTheme(
      <Button size="small">Click me</Button>
    )
    expect(getByText('Click me')).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })
  it('3. should render the small medium by default', () => {
    const { getByText } = renderWithTheme(<Button>Click me</Button>)
    expect(getByText('Click me')).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem'
    })
  })
  it('4. should render the large size', () => {
    const { getByText } = renderWithTheme(
      <Button size="large">Click me</Button>
    )
    expect(getByText('Click me')).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem'
    })
  })
  it('5. should render the button with full width', () => {
    const { getByText } = renderWithTheme(<Button fullWidth>Click me</Button>)
    expect(getByText('Click me')).toHaveStyle('width: 100%')
  })
  it('6. should call onClick when clicked', () => {
    const onClick = jest.fn()
    const { getByText } = renderWithTheme(
      <Button onClick={onClick}>Click me</Button>
    )
    fireEvent.click(getByText('Click me'))
    expect(onClick).toHaveBeenCalled()
  })
  it('7. should render Button as a Link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Click me
      </Button>
    )

    expect(screen.getByRole('link', { name: /Click me/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
