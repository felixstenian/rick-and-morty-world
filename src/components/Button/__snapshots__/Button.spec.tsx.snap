// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> 1. should render the reading 1`] = `
.c0 {
  border: 0;
  cursor: pointer;
  width: 20rem;
  background: linear-gradient(180deg,#BFDE42 0%,#F5EB46 100%);
  color: #2E2F42;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
}

.c0:hover {
  background: #8DC760;
}

<button
  class="c0"
>
  Click me
</button>
`;
